File: client/jetbrains/src/main/java/com/sourcegraph/cody/ui/AccordionSection.java
package com.sourcegraph.cody.ui;

import com.intellij.openapi.ui.VerticalFlowLayout;
import java.awt.*;
import javax.swing.*;

public class AccordionSection extends JPanel {
  private final JButton toggleButton;
  private final String sectionTitle;
  private JPanel contentPanel;

  public AccordionSection(String title) {
    setLayout(new BorderLayout());
    sectionTitle = title;

    toggleButton = new JButton(createToggleButtonHTML(title, true));
    toggleButton.setHorizontalAlignment(SwingConstants.LEFT);
    toggleButton.setBorderPainted(false);
    toggleButton.setFocusPainted(false);
    toggleButton.setContentAreaFilled(false);
    toggleButton.addActionListener(
        e -> {
          if (contentPanel.isVisible()) {
            contentPanel.setVisible(false);
            toggleButton.setText(createToggleButtonHTML(sectionTitle, true));
          } else {
            contentPanel.setVisible(true);
            toggleButton.setText(createToggleButtonHTML(sectionTitle, false));
          }
        });

    contentPanel = new JPanel();
    contentPanel.setLayout(new VerticalFlowLayout(VerticalFlowLayout.TOP, 0, 0, true, false));
    contentPanel.setVisible(false);

    add(toggleButton, BorderLayout.NORTH);
    add(contentPanel, BorderLayout.CENTER);
  }

  private String createToggleButtonHTML(String title, boolean isCollapsed) {
    String symbol =
        isCollapsed ? "&#9658;" : "&#9660;"; // Unicode entities for right and down arrows
    return "<html><body style='text-align:left'>"
        + title
        + " <span style='float:right; color:gray'>"
        + symbol
        + "</span></body></html>";
  }

  public JPanel getContentPanel() {
    return contentPanel;
  }
}

File: client/jetbrains/src/main/kotlin/com/sourcegraph/cody/ui/AccordionSection.kt
package com.sourcegraph.cody.ui

import com.intellij.openapi.ui.VerticalFlowLayout
import java.awt.BorderLayout
import javax.swing.JButton
import javax.swing.JPanel
import javax.swing.SwingConstants

class AccordionSection(title: String) : JPanel() {
  val contentPanel: JPanel
  private val toggleButton: JButton
  private val sectionTitle: String

  init {
    layout = BorderLayout()
    sectionTitle = title
    toggleButton = JButton(createToggleButtonHTML(title, true))
    toggleButton.horizontalAlignment = SwingConstants.LEFT
    toggleButton.isBorderPainted = false
    toggleButton.isFocusPainted = false
    toggleButton.isContentAreaFilled = false
    contentPanel = JPanel()
    toggleButton.addActionListener { _ ->
      if (contentPanel.isVisible) {
        contentPanel.isVisible = false
        toggleButton.text = createToggleButtonHTML(sectionTitle, true)
      } else {
        contentPanel.isVisible = true
        toggleButton.text = createToggleButtonHTML(sectionTitle, false)
      }
    }
    contentPanel.layout = VerticalFlowLayout(VerticalFlowLayout.TOP, 0, 0, true, false)
    contentPanel.isVisible = false
    add(toggleButton, BorderLayout.NORTH)
    add(contentPanel, BorderLayout.CENTER)
  }

  private fun createToggleButtonHTML(title: String, isCollapsed: Boolean): String {
    val symbol =
        if (isCollapsed) "&#9658;" else "&#9660;" // Unicode entities for right and down arrows
    return ("<html><body style='text-align:left'>" +
        title +
        " <span style='float:right; color:gray'>" +
        symbol +
        "</span></body></html>")
  }
}
